### PR TITLE
[vSphere][datastore] Add support for additional metrics

### PR DESCRIPTION
**Pull request by lucian-ioan**
## Proposed commit message

Following beats code changes in https://github.com/elastic/beats/pull/40441, the following metrics need integration update:
Metrics Type | Metrics | Field Mappings
-- | -- | -- |
Datastore performance |   |   |
| | datastore.read.average  (KBps) | datastore.read.bytes |
| | datastore.write.average (KBps) | datastore.write.bytes |
| | datastore.datastoreIops.average (num) | datastore.iops |
| | datastore.totalReadLatency.average (ms) | datastore.read.latency.total.ms |
| | datastore.totalWriteLatency.average (ms) | datastore.write.latency.total.ms |
Datastore Summry|   |   |
| | Datastore.OverallStatus | datastore.status |
| | Datastore.Host | datastore.host.count |
| | Datastore.Host | datastore.host.names |
| | Datastore.Vm | datastore.vm.count |
| | Datastore.Vm | datastore.vm.names |

</body>
</html>

## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [ ] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [ ] 

## Related issues

-  Closes https://github.com/elastic/obs-integration-team/issues/29

### CODE REVIEW

File: packages/vsphere/changelog.yml [hunk 1]
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "1.14.0"
```
Comment (ishleenk17) [inline packages/vsphere/changelog.yml:2]: Until all changes are ready to be merged, lets keep this as "1.14.0-next".
Also for now we will not be incrementing the version in manifest. 

Once we are ready to push all the changes, we will bump the kibana version in manifest.

**Comment by lalit-satapathy** [comment]
What about stack version dependancy?

Also release version we need to use something else?

**Comment by ishleenk17** [comment]
I have shared the details of how we can handle versioning [here](https://github.com/elastic/obs-integration-team/issues/22).
Please follow that for all the Integration PR's

**Comment by ishleenk17** [comment]
@agithomas : Keeping TSDB in mind, can you check and suggest if any of these need to be added for TSDB?

File: packages/vsphere/data_stream/datastore/fields/fields.yml [hunk 2]
```diff
@@ -6,33 +6,85 @@
...
       format: percent
+    - name: host.count
+      type: long
+      description: >
+        Number of hosts.
+
+    - name: host.names
+      type: keyword
+      description: >
+        List of all the host names.
```
[the remark this answers is not in the document]
Comment (lucian-ioan) [reply packages/vsphere/data_stream/datastore/fields/fields.yml:47]: @agithomas Could we use the fingerprint processor or do you think it's unnecessary?

**Comment by lucian-ioan** [comment]
@agithomas @ishleenk17 I think host.names and vm.names could be used for TSDB.

Do you think using the fingerprint processor is unnecessary here?

File: packages/vsphere/data_stream/datastore/fields/fields.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ishleenk17) [reply packages/vsphere/data_stream/datastore/fields/fields.yml]: I would recommend against it. Since we have seen issues with the processor. 
Also, I assume now ES supports this functionality.
Agi can confirm this.

File: packages/vsphere/data_stream/datastore/fields/fields.yml [hunk 3]
```diff
@@ -6,33 +6,75 @@
...
       metric_type: gauge
       unit: percent
-      description: Used percent of the datastore
+      description: Percentage of datastore capacity used.
       format: percent
+    - name: host.count
+      type: long
+      description: >
+        Number of hosts associated with the datastore.
+    - name: host.names
```
Comment (agithomas) [inline packages/vsphere/data_stream/datastore/fields/fields.yml:43]: I strongly believe, `service.address`, fingerprint applied on `host.name` atleast is needed as a dimension field

Comment (agithomas) [reply packages/vsphere/data_stream/datastore/fields/fields.yml:43]: Can there be a possibility, where `service.address` is not unique, in a multi-region, multi-cluster setup?

Comment (agithomas) [reply packages/vsphere/data_stream/datastore/fields/fields.yml:43]: Any reason why `agent.id` is excluded as a dimension filed?

**Comment by agithomas** [comment]
> @agithomas @ishleenk17 I think `host.names` and `vm.names` could be used for TSDB.
> 
> Do you think using the fingerprint processor is unnecessary here?

Going with the description of `host.names`, it is `List of all the host names associated with the datastore.`. So, can there be more than one hosts mapped to one datastore? So, is this an array?  Does dimensions support arrays?

File: packages/vsphere/data_stream/datastore/fields/fields.yml (hunk 3 above)
Comment (ishleenk17) [reply packages/vsphere/data_stream/datastore/fields/fields.yml:43]: @agithomas : IIRC, usage of fingerprint was discouraged and we planned to revert all the the usage in Integrations.
Or was it specific to labels fingerprint ?

Comment (agithomas) [reply packages/vsphere/data_stream/datastore/fields/fields.yml:43]: If i am not wrong, there are two different recent recommendations. 

1. Do not create a fingerprint even though you have exceeded the dimension limits.
2. Locality-sensitive hashing with which you can mark labels.* as `dimension:true`

Here, i guess,  option 2, `host.*` can be attempted.

Comment (agithomas) [reply packages/vsphere/data_stream/datastore/fields/fields.yml:43]: Kindly refer to the [PR](https://github.com/elastic/integrations/pull/9785/) for more details. Unlike Prometheus, there is no problem with Vsphere integration as this is a new integration dataset

**Comment by ishleenk17** [comment]
> > @agithomas @ishleenk17 I think `host.names` and `vm.names` could be used for TSDB.
> > Do you think using the fingerprint processor is unnecessary here?
> 
> Going with the description of `host.names`, it is `List of all the host names associated with the datastore.`. So, can there be more than one hosts mapped to one datastore? So, is this an array? Does dimension mapping support arrays?

Yes, one datastore can have multiple hosts associated.

File: packages/vsphere/data_stream/datastore/fields/fields.yml (hunk 3 above)
Comment (ishleenk17) [reply packages/vsphere/data_stream/datastore/fields/fields.yml:43]: > Any reason why `agent.id` is excluded as a dimension filed?

Its not excluded. ITs present if you see the actual code. 
https://github.com/elastic/integrations/blob/main/packages/vsphere/data_stream/datastore/fields/agent.yml

Comment (agithomas) [reply packages/vsphere/data_stream/datastore/fields/fields.yml:43]: Apologies, i thought this is a new integration and missed to check the existing mappings.

Comment (agithomas) [reply packages/vsphere/data_stream/datastore/fields/fields.yml:43]: > Can there be a possibility, where service.address is not unique, in a multi-region, multi-cluster setup?
Already there exist dimension mapping. Please ignore this comment

**Comment by agithomas** [comment]
@ishleenk17 , if a specific datastore is mapped with more than one hosts, will it create multiple documents or it will be a single document having `host.names` having multiple values?

This decides - if `host.names` must be a dimension or not

**Comment by ishleenk17** [comment]
> @ishleenk17 , if a specific datastore is mapped with more than one hosts, will it create multiple documents or it will be a single document having `host.names` having multiple values?
> 
> This decides - if `host.names` must be a dimension or not

Single document with host.name/vm.name as an array in it. [Sample json](https://github.com/elastic/integrations/pull/10911/files#diff-51647c4010743d09e3a1ed346b052733ae3ac27250320a3f6f1571f991cc70e4R88)

**Comment by agithomas** [comment]
> Single document with host.name/vm.name as an array in it. [Sample json](https://github.com/elastic/integrations/pull/10911/files#diff-51647c4010743d09e3a1ed346b052733ae3ac27250320a3f6f1571f991cc70e4R88)

Thanks for clarifying. 

I guess, there is enough dimensions for this dataset. 

Recommend to test with TSDB test toolkit for verification to avoid missing any edge cases. 

The PR looks good from TSDB point of view.

File: packages/vsphere/data_stream/datastore/fields/fields.yml [hunk 4]
```diff
@@ -6,33 +6,75 @@
...
+      type: keyword
+      description: >
+        List of all the host names associated with the datastore.
+    - name: iops
+      type: long
+      description: >
+        Storage I/O Control aggregated Input/Output Operations Per Second.
+    - name: read.bytes
+      type: long
+      description: >
```
Comment (ishleenk17) [inline packages/vsphere/data_stream/datastore/fields/fields.yml:53]: metric_type to be added for new metrics.

**Comment by lucian-ioan** [comment]
@ishleenk17 From what I understood, we do not need any more fields for TSDB.

TSDB was already enabled previously and I see it still works fine with the changes:
image

Looks like this PR is good to go.

File: packages/vsphere/data_stream/datastore/fields/fields.yml [hunk 5]
```diff
@@ -6,33 +6,86 @@
...
+        List of all the VM names associated with the datastore.
+    - name: write.bytes
+      type: long
+      metric_type: gauge
+      unit: byte
+      description: >
+        Rate of writing data to the datastore.
+      format: bytes
+    - name: write.latency.total.ms
```
Comment (ishleenk17) [inline packages/vsphere/data_stream/datastore/fields/fields.yml:86]: Please check the metricset of datastore. There is a change in the fields. 
Lets get the latest fields in.

**Comment by ishleenk17** [comment]
I don't see the period setting present here per datastream. 
Is that uet to be added ?
I thought you were testing with those changes @lucian-ioan

**Comment by lucian-ioan** [comment]
Updated the integration with all the recent changes from elastic/beats.

File: packages/vsphere/changelog.yml [hunk 6]
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "1.14.0-next"
```
Comment (ishleenk17) [inline packages/vsphere/changelog.yml:2]: Update the version to 1.15.0-next

File: packages/vsphere/data_stream/datastore/fields/fields.yml (hunk 5 above)
Comment (ishleenk17) [reply packages/vsphere/data_stream/datastore/fields/fields.yml:109]: I don't think this is resolved still ?

**Comment by ishleenk17** [comment]
Period is not removed from the main manifest file and upgrade the kibana version to 8.15.2

File: packages/vsphere/data_stream/datastore/manifest.yml [hunk 7]
```diff
@@ -5,6 +5,19 @@ streams:
     title: vSphere datastore metrics
     description: vSphere datastore metrics
     vars:
+      - name: period
+        type: text
+        title: Period
+        description: The vSphere performance API allows for collecting host and datastore metrics at various intervals, including real-time (every 20 seconds), 5 minutes, or longer durations. Properly configuring the "period" parameter is crucial to manage usage restrictions and ensure accurate data collection. For more information on optimizing your data collection setup, refer to the [Data Collection Intervals](https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.monitoring.doc/GUID-247646EA-A04B-411A-8DD4-62A3DCFCF49B.html) and [Data Collection Levels](https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.monitoring.doc/GUID-25800DE4-68E5-41CC-82D9-8811E27924BC.html) documentation.
+        default: 20s
+      - name: tags
```
Comment (ishleenk17) [inline packages/vsphere/data_stream/datastore/manifest.yml:13]: Tags is not added to hbs file.

**Comment by ishleenk17** [comment]
Can we add pipeline tests as well ?
You may refer to host PR

**Comment by ishleenk17** [comment]
The fields in the commit description are not updated to the latest ones

**Review by niraj-elastic — APPROVED** [review]
LGTM!

**Review by kush-elastic — APPROVED** [review]
LGTM!

**Review by ishleenk17 — APPROVED** [review]
Looks good!
